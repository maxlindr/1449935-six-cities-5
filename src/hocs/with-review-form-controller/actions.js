export const ActionType = {
  SET_RATING: `SET_RATING`,
  SET_TEXT: `SET_TEXT`,
  SET_VALID: `SET_VALID`,
  SET_PENDING: `SET_PENDING`,
  VALIDATE: `VALIDATE`,
  RESET: `RESET`,
};

export const ActionCreator = {
  setRating: (rating) => ({
    type: ActionType.SET_RATING,
    payload: rating
  }),
  setText: (text) => ({
    type: ActionType.SET_TEXT,
    payload: text
  }),
  setPending: (value) => ({
    type: ActionType.SET_PENDING,
    payload: value
  }),
  validate: () => ({
    type: ActionType.VALIDATE
  }),
  reset: () => ({
    type: ActionType.RESET
  }),
};
