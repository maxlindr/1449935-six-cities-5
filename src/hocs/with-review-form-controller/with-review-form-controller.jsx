import React from 'react';
import PropTypes from 'prop-types';
import {connect} from 'react-redux';
import {postComment} from '../../store/api-actions';

const INITIAL_STATE = {
  rating: null,
  text: ``,
  isValid: false,
  isDisabled: false,
};

const withReviewFormController = (Component) => {
  class WithReviewFormController extends React.PureComponent {
    constructor(props) {
      super(props);

      this.state = INITIAL_STATE;

      this.handleSubmit = this.handleSubmit.bind(this);
      this.handleRatingClick = this.handleRatingClick.bind(this);
      this.handleTextChange = this.handleTextChange.bind(this);
    }

    handleSubmit(evt) {
      evt.preventDefault();

      const {offerId, dispatch} = this.props;

      const comment = {
        text: this.state.text,
        rating: this.state.rating
      };

      const submitComment = () => {
        dispatch(postComment(offerId, comment))
          .then(() => this.setState(INITIAL_STATE))
          .catch(() => this.setState({
            isDisabled: false
          }));
      };

      this.setState({isDisabled: true}, submitComment);
    }

    handleRatingClick(evt) {
      this.setState({rating: Number(evt.target.value)}, this.validate);
    }

    handleTextChange(evt) {
      this.setState({text: evt.target.value}, this.validate);
    }

    validate() {
      const {rating, text} = this.state;

      this.setState({isValid: Boolean(text.trim() && rating)});
    }

    render() {
      const {text, rating, isValid, isDisabled} = this.state;

      return (
        <Component
          text={text}
          rating={rating}
          isValid={isValid}
          disabled={isDisabled}
          onTextChange={this.handleTextChange}
          onRatingClick={this.handleRatingClick}
          onSubmit={this.handleSubmit}
        />
      );
    }
  }

  WithReviewFormController.propTypes = {
    offerId: PropTypes.string.isRequired,
    dispatch: PropTypes.func.isRequired,
  };

  return connect()(WithReviewFormController);
};

export default withReviewFormController;
