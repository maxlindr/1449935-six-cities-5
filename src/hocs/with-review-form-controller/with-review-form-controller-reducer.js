import {ActionType} from './actions';

const MIN_COMMENT_LENGTH = 50;
const MAX_COMMENT_LENGTH = 300;

export const INITIAL_STATE = {
  rating: null,
  text: ``,
  isValid: false,
  isPending: false,
};

const checkCommentValidity = (text) => {
  const trimmedText = text.trim();

  return (
    trimmedText.length >= MIN_COMMENT_LENGTH &&
    trimmedText.length <= MAX_COMMENT_LENGTH
  );
};

export default (state, action) => {
  switch (action.type) {
    case ActionType.SET_RATING:
      return Object.assign({}, state, {
        rating: action.payload
      });
    case ActionType.SET_TEXT:
      return Object.assign({}, state, {
        text: action.payload
      });
    case ActionType.SET_VALID:
      return Object.assign({}, state, {
        isValid: action.payload
      });
    case ActionType.SET_PENDING:
      return Object.assign({}, state, {
        isPending: action.payload
      });
    case ActionType.RESET:
      return INITIAL_STATE;
    case ActionType.VALIDATE:
      return Object.assign({}, state, {
        isValid: checkCommentValidity(state.text) && state.rating > 0
      });
    default:
      return state;
  }
};
