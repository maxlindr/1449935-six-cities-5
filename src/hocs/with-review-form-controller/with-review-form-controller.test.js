import React from 'react';
import renderer from 'react-test-renderer';
import PropTypes from 'prop-types';
import {withReviewFormController} from './with-review-form-controller';

const MockComponent = () => <div>Mock Component</div>;

MockComponent.propTypes = {
  text: PropTypes.string.isRequired,
  rating: PropTypes.number,
  isValid: PropTypes.bool.isRequired,
  disabled: PropTypes.bool.isRequired,
  onTextChange: PropTypes.func.isRequired,
  onRatingClick: PropTypes.func.isRequired,
  onSubmit: PropTypes.func.isRequired,
};

const MockComponentWrapped = withReviewFormController(MockComponent);

test(`withReviewFormController должен корректно отображаться`, () => {
  const tree = renderer.create((
    <MockComponentWrapped
      offerId={`1`}
      postComment={() => {}}
    />
  )).toJSON();

  expect(tree).toMatchSnapshot();
});
