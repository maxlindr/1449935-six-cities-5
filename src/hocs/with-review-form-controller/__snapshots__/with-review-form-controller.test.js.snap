// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`withReviewFormController должен корректно отображаться 1`] = `
<div>
  Mock Component
</div>
`;
